pkg/utils/gardener/cloudprofile.go

```diff
@@ -412,3 +414,117 @@ func NewV1beta1ImagesContext(parentImages []gardencorev1beta1.MachineImage) *Ima
+// It returns a list of field errors if any validation fails.
+func ValidateCapabilities(capabilities core.Capabilities, capabilitiesDefinitions []core.CapabilityDefinition, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// create map from capabilitiesDefinitions
+	capabilitiesDefinition := make(map[string][]string)
+	for _, capabilityDefinition := range capabilitiesDefinitions {
+		capabilitiesDefinition[capabilityDefinition.Name] = capabilityDefinition.Values
+	}
+	supportedCapabilityKeys := slices.Collect(maps.Keys(capabilitiesDefinition))
+
+	// Check if all capabilities are supported by the cloud profile
+	for capabilityKey, capability := range capabilities {
+		supportedValues, keyExists := capabilitiesDefinition[capabilityKey]
+		if !keyExists {
+			allErrs = append(allErrs, field.NotSupported(fldPath, capabilityKey, supportedCapabilityKeys))
+			continue
+		}
+		for i, value := range capability {
+			if !slices.Contains(supportedValues, value) {
+				allErrs = append(allErrs, field.NotSupported(fldPath.Child(capabilityKey).Index(i), value, supportedValues))
+			}
+		}
+	}
+
+	// Check additional requirements for architecture
+	// must be defined when multiple architectures are supported by the cloud profile
+	supportedArchitectures := capabilitiesDefinition[v1beta1constants.ArchitectureName]
+	architectures := capabilities[v1beta1constants.ArchitectureName]
+	if len(supportedArchitectures) > 1 && len(architectures) != 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child(v1beta1constants.ArchitectureName), architectures, "must define exactly one architecture when multiple architectures are supported by the cloud profile"))
+	}
+
+	return allErrs
+}
+
+// GetVersionCapabilitySets returns the CapabilitySets for a given machine image version or adds the default capabilitySet if none is defined.
+func GetVersionCapabilitySets(version core.MachineImageVersion, capabilitiesDefinitions []core.CapabilityDefinition) []core.CapabilitySet {
+	versionCapabilitySets := version.CapabilitySets
+	if len(versionCapabilitySets) != 0 {
+		return versionCapabilitySets
+	}
+
+	// find architecture Capability in capabilitiesDefinitions
+	// the architecture capability must exist in every cloud profile
+	var supportedArchitectures []string
+	for _, capabilityDefinition := range capabilitiesDefinitions {
+		if capabilityDefinition.Name == v1beta1constants.ArchitectureName {
+			supportedArchitectures = capabilityDefinition.Values
+			break
+		}
+	}
+
+	// It is allowed not to define capabilitySets in the machine image version if there is only one architecture
+	// if so the capabilityDefinitions are used as default
+	if len(supportedArchitectures) == 1 {
+		capabilities := make(core.Capabilities)
+		versionCapabilitySets = []core.CapabilitySet{{Capabilities: ApplyDefaultCapabilities(capabilities, capabilitiesDefinitions)}}
```

Comment:
```suggestion
		versionCapabilitySets = []core.CapabilitySet{{Capabilities: ApplyDefaultCapabilities(nil, capabilitiesDefinitions)}}
```
To me it seems that the capabilities parameter may be nil and will be initialized in the function.